File: AlgorithmsPackCPP/91. Decode Ways.cpp
const int MAX = 100+1;
int memory[MAX];
string str;

class Solution {
public:
    bool can2Digit(int index){
        if(index+1 >=str.size())
            return false;
        if(str[index]=='1')
            return true;
        return str[index]=='2' && str[index+1]<='6';
    }
    
    int countWays(int index){
        if(index==str.size())
            return 1;
        if(str[index]=='0')
            return 0;
        auto&ret = memory[index];
        if(ret!=-1)
            return ret;
        ret = countWays(index+1);
        if(can2Digit(index))
            ret+=countWays(index+2);
        return ret;
    }
    
    int numDecodings(string s) {
        str = s;
        memset(memory,-1,sizeof(memory));
        return countWays(0);
    }
};

File: AlgorithmsPackPythonSolutions/91. Decode Ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        MAX = 100 + 1
        memory = [-1] * MAX
        str_length = len(s)

        def can_2_digit(index):
            if index + 1 >= str_length:
                return False
            if s[index] == '1':
                return True
            return s[index] == '2' and s[index + 1] <= '6'

        def count_ways(index):
            if index == str_length:
                return 1
            if s[index] == '0':
                return 0
            if memory[index] != -1:
                return memory[index]
            result = count_ways(index + 1)
            if can_2_digit(index):
                result += count_ways(index + 2)
            memory[index] = result
            return result

        return count_ways(0)
